receiver/awscontainerinsightreceiver/internal/k8sapiserver/metric_scrape.go

```diff
@@ -0,0 +1,171 @@
+	// TODO
+	/* 2023-05-16T18:00:47.956Z        debug   scrape/scrape.go:1351   Scrape failed   {"kind": "receiver", "name": "awscontainerinsightreceiver", "data_type": "metrics", "scrape_pool": "k8sapiserver", "target": "https://ip-192-168-71-146.us-east-2.compute.internal:443/metrics", "error": "Get \"https://ip-192-168-71-146.us-east-2.compute.internal:443/metrics\": unable to read authorization credentials file /var/Run/secrets/kubernetes.io/serviceaccount/token: open /var/Run/secrets/kubernetes.io/serviceaccount/token: no such file or directory"}
+	 */
+
+	config := &config.Config{
+		ScrapeConfigs: []*config.ScrapeConfig{
+			{
+				JobName: "k8sapiserver",
+
+				ScrapeInterval: model.Duration(collectionInterval),
+				ScrapeTimeout:  model.Duration(collectionTimeout),
+
+				HonorLabels:     true,
+				HonorTimestamps: true,
+
+				MetricsPath: collectionEndpoint,
+				Scheme:      "https",
+
+				HTTPClientConfig: httpconfig.HTTPClientConfig{
+					TLSConfig: httpconfig.TLSConfig{
+						InsecureSkipVerify: true,
+					},
+					BearerTokenFile: bearerTokenPath,
+					FollowRedirects: true,
+				},
+
+				ServiceDiscoveryConfigs: discovery.Configs{
+					discovery.StaticConfig{
+						&targetgroup.Group{
+							Targets: []model.LabelSet{
+								{model.AddressLabel: model.LabelValue(host)},
+							},
+							Source: "0",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// the prometheus scraper library uses gokit logging, we must adapt our zap logger to gokit
+	gokitlogger := logging.NewZapToGokitLogAdapter(logger)
+
+	ms := MetricScrape{
+		logger:      logger,
+		gokitlogger: gokitlogger,
+		store:       store,
+		config:      config,
+	}
+
+	ms.context, ms.cancel = context.WithCancel(context.Background())
+
+	ms.scrapeManager = scrape.NewManager(&scrape.Options{PassMetadataInContext: true}, gokitlogger, *store)
+	ms.discoveryManager = discovery.NewManager(ms.context, gokitlogger)
+
+	return &ms, nil
+}
+
+func (ms *MetricScrape) Run() error {
+	err := ms.scrapeManager.ApplyConfig(ms.config)
+	if err != nil {
+		return err
+	}
+
+	discoveryConfigs := make(map[string]discovery.Configs)
+	discoveryConfigs["k8sapiserver"] = ms.config.ScrapeConfigs[0].ServiceDiscoveryConfigs
+	err = ms.discoveryManager.ApplyConfig(discoveryConfigs)
+	if err != nil {
+		return err
+	}
+
+	go ms.runDiscoveryManager()
+	go ms.runScrapeManager()
+
+	return nil
+}
+
+func (ms *MetricScrape) Shutdown(context.Context) error {
+	if ms.cancel != nil {
+		ms.cancel()
+	}
+	if ms.scrapeManager != nil {
+		ms.scrapeManager.Stop()
+	}
+	return nil
+}
+
+func (ms *MetricScrape) runDiscoveryManager() {
+	ms.logger.Info("Starting k8sapiserver prometheus metrics discovery manager")
+	if err := ms.discoveryManager.Run(); err != nil {
+		ms.logger.Error("Discovery manager failed", zap.Error(err))
+		return
+	}
+
+	select {
+	case <-ms.context.Done():
+		ms.logger.Info("Stopping k8sapiserver prometheus metrics discovery manager")
+		return
```

Comment:
nit - Don't think this return is necessary. It's going to return after logging anyway, pretty sure

Reply:
sounds good, yeah not sure which style we prefer, implicit or explicit